lessons/gitBranches/lesson.md

```diff
@@ -0,0 +1,162 @@
+any individual commit can have multiple 'children'. This history can be traced
+back through the 'lineage' or 'ancestry'. The branch just gives a name to each
+'lineage' when a commit has multiple children.
+
+When you merge two branches together, the commit histories get merged together
+as well. Which means that all the changes you made in each branch gets combined
+back into a single lineage, rather than two. This makes it easier to work
+collaboratively on a project, since each individual could work on their own
+branches, without dealing with the messiness that could come from working all on
+one branch.
+
+## Commands used with branches
+
+Branches are best understood visually. So let's first start with using 
+[this website](https://onlywei.github.io/explain-git-with-d3/) to see what the 
+`branch`, `checkout`, and `merge` commands are doing.
+
+After we've tried that, let's do it locally (on your own computer). Here is a
+sequence of commands to try out:
+
+    cd ~/Desktop
+    mkdir git-branches
+    cd git-branches
+    git init # start a repo
+    git add .
+    git commit -m "First commit" # make the first commit
+    git branch testBranch
+    git checkout testBranch # create and move to a new branch
+    ## can also do git checkout -b testBranch
+    echo "Some text" > file.txt 
+    git add file.txt
+    git commit -m "Added a file with text"
+    git checkout master
+    echo "Text in another file" > new-file.txt
+    git add new-file.txt
+    git commit -m "Added another file"
+    git log --graph --oneline
+    git merge testBranch
+    git log --graph --oneline
+    git branch -d testBranch # delete the branch
+    
+## Using branches for pull requests 
+
+I mentioned this already, but branches are best used when doing a 
+[pull request](https://help.github.com/articles/using-pull-requests/)
+(unless the pull request is very small or few people work on the repository).
+
+The steps to take would be:
+
+1. Fork a repository on GitHub
+2. Clone it onto your computer
+3. Make a branch and move to it: `git checkout -b fixingBranch`
+4. Make changes to the files
+5. Commit the changes to the history
+6. Push the branch up to your forked version: `git push origin fixingBranch`
+7. On GitHub, submit a Pull Request of your `fixingBranch`
+8. Once the pull request is merged, 
+[delete](https://github.com/blog/1377-create-and-delete-branches) 
+the `fixingBranch` on your forked repo on GitHub and on your computer 
+(`git pull upstream master && git branch -d fixingBranch`)
+
+# Resources: #
+
+If you have any questions, often one of the best places to start is either  
+[StackOverflow](http://stackoverflow.com/questions/tagged/git) or Google (which
+more likely links to StackOverflow).
+
+## Glossary: ##
+
+* `cd` - change directory
+* directory - the same thing as a folder
+* `mkdir` - make a directory
+* `echo` - print a message to the screen or to a file if `>` (redirect) is
+present.
+* `init` - start or initialize a git repository
```

Comment:
This and the following commands are all subcommands to git, which is confusing since the 4 above are not.